src/libraries/System.Resources.ResourceManager/tests/ResourceManagerTests.cs

```diff
@@ -222,6 +225,46 @@ public static void IgnoreCase(string key, string expectedValue)
             Assert.Equal(expectedValue, manager.GetString(key.ToLower(), culture));
         }
 
+        /// <summary>
+        /// This test has multiple threads simultaneously loop over the keys of a moderately-sized resx using
+        /// <see cref="ResourceManager"/> and call <see cref="ResourceManager.GetString(string)"/> for each key.
+        /// This has historically been prone to thread-safety bugs because of the shared cache state and internal
+        /// method calls between RuntimeResourceSet and <see cref="ResourceReader"/>.
+        ///
+        /// Running with <paramref name="useEnumeratorEntry"/> TRUE replicates https://github.com/dotnet/runtime/issues/74868,
+        /// while running with FALSE replicates the error from https://github.com/dotnet/runtime/issues/74052.
+        /// </summary>
+        /// <param name="useEnumeratorEntry">
+        /// Whether to use <see cref="IDictionaryEnumerator.Entry"/> vs. <see cref="IDictionaryEnumerator.Key"/> when enumerating;
+        /// these follow fairly different code paths.
+        /// </param>
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public static void TestResourceManagerIsSafeForConcurrentAccessAndEnumeration(bool useEnumeratorEntry)
+        {
+            ResourceManager manager = new("System.Resources.Tests.Resources.AToZResx", typeof(ResourceManagerTests).GetTypeInfo().Assembly);
+
+            const int Threads = 10;
+            using Barrier barrier = new(Threads);
+            Task task = Task.WhenAll(Enumerable.Range(0, Threads).Select(_ => Task.Run(WaitForBarrierThenEnumerateResources)));
+
+            Assert.True(task.Wait(TimeSpan.FromSeconds(10)));
```

Comment:
Before my fixes, this test reliably would reproduce both issues on my machine. Let me know if there is a better way to test for concurrency issues like this. 

I do think it is very important to have such a test, because this problem was actually previously [fixed in 2015](https://github.com/dotnet/coreclr/pull/75/files) and then broken again in 2020 (I think the break happened [here](https://github.com/dotnet/runtime/pull/44454/files#diff-21262dbd9abd6640e1b60d773746e46f6bb796efc8c4ddab9aa9be6f975b284bL318).